### PR TITLE
State that `Evolution` has matrix if target is `Hamiltonian`

### PR DESCRIPTION
**Pull request by albi3ro**
We discovered a difference between `default.qubit` and `default.qubit.legacy` over their treatment of the evolution of hamiltonians:

```python
dev = qml.device('default.qubit', wires=3)
# dev = qml.device('default.qubit.legacy', wires=3)  # this works!
​
coeffs = [0.1, 0.2]
operators = [qml.PauliZ(0) @ qml.PauliZ(1), qml.PauliZ(0) @ qml.PauliZ(2)]
hamiltonian = qml.Hamiltonian(coeffs, operators)
​
@qml.qnode(dev)
def f():
    qml.evolve(hamiltonian, 0.5)
    return qml.density_matrix([0])
​
f()
```

Previously `qml.evolve(hamiltonain, t)` worked because `Evolution` was hardcoded as an accepted observable. This meant it became an accepted operation as well, even though it stated it didn't have a matrix.

In `ScalarSymbolicOp`, the parent for `Evolution`, we added a special logical pathway for the matrix if the target is a Hamiltonian, but we didn't include that special logical pathway for the `has_matrix` property.

This PR updates the `has_matrix` property for and allows the evolution of a hamiltonian to execute on `default.qubit`.

**Warning:**

This matrix is not backprop compatible.

```
dev = qml.device('default.qubit.legacy', wires=3)

coeffs = [0.1, 0.2]
operators = [qml.PauliZ(0) @ qml.PauliZ(1), qml.PauliZ(0) @ qml.PauliZ(2)]

@qml.qnode(dev)
def f(x):
    hamiltonian = qml.Hamiltonian(x, operators)
    qml.evolve(hamiltonian, 0.5)
    return qml.density_matrix([0])

qml.grad(f)(qml.numpy.array(coeffs))
```
Gives
```
AttributeError: 'ArrayBox' object has no attribute 'tocsr'
```

`qml.dot` and operator arithemetic does not have these problems.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -33,12 +33,16 @@
 
 <h3>Bug fixes 🐛</h3>
 
+* Any `ScalarSymbolicOp`, like `Evolution`, now states that it has a matrix if the target
+  is a `Hamiltonian`.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:37]:
```suggestion
  is a `Hamiltonian`.
  [(#4768)](https://github.com/PennyLaneAI/pennylane/pull/4768)
```

File: tests/ops/op_math/test_symbolic_op.py [hunk 2]
```diff
@@ -139,6 +139,13 @@ class DummyOp(Operator):
         op = SymbolicOp(base)
         assert op.has_matrix == has_mat
 
+    def test_has_matrix_hamiltoanin(self):
```
Comment (lillian542) [inline tests/ops/op_math/test_symbolic_op.py:142]:
```suggestion
    def test_has_matrix_hamiltonian(self):
```

**Review by lillian542 — APPROVED** [review]
Looks good to me! 🎉

**Review by mudit2812 — APPROVED** [review]

**Review by timmysilv — APPROVED** [review]

**Comment by albi3ro** [comment]
[sc-50818]

**Pull request merged**